akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala

```diff
@@ -344,6 +346,53 @@ class ClientServerSpec extends WordSpec with Matchers with BeforeAndAfterAll wit
         }(system2)
         Await.result(b1.unbind(), 1.second)
       }(materializer2)
+
+      "stop stages on failure" in Utils.assertAllStagesStopped {
+        val (_, hostname, port) = TestUtils.temporaryServerHostnameAndPort()
+        val counter = new AtomicLong(0)
+        val stage: GraphStage[FlowShape[HttpRequest, HttpResponse]] = new GraphStage[FlowShape[HttpRequest, HttpResponse]] {
+          val in = Inlet[HttpRequest]("request.in")
+          val out = Outlet[HttpResponse]("response.out")
+
+          override def shape: FlowShape[HttpRequest, HttpResponse] = FlowShape.of(in, out)
+
+          override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler with OutHandler {
+            override def preStart(): Unit = counter.incrementAndGet()
+            override def postStop(): Unit = counter.decrementAndGet()
+            override def onPush(): Unit = push(out, HttpResponse(entity = counter.get().toString))
+            override def onPull(): Unit = pull(in)
```

Comment:
maybe overwrite onUpstreamCompleted as well and count there in addition to the existing counter here?